### PR TITLE
Blog App - processing data in models

### PR DESCRIPTION
**Pull request by batoolfatima2135**
In this branch, I have: 
### Created models
-  Created model classes for all entities as shown in the [ERD diagram](../images/blog_app_erd_v1_1.png).
-  Set up associations between models.

### Used models to insert data
- Created users 
- Created  posts written by one of the users
- Created comments for one of the posts created
- Use other [CRUD methods](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/lessons/active_record_crud.md) to find, update and delete entities.

### Created custom methods
models include some additional methods.
-  Users
    - A method that returns the 3 most recent posts for a given user.
-  Posts
    - A method that updates the posts counter for a user.
    - A method which returns the 5 most recent comments for a given post.
-  Comments
    - A method that updates the comments counter for a post.
-  Likes
    - A method that updates the likes counter for a post.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -1,2 +1,8 @@
 class User < ApplicationRecord
+  has_many :posts, foreign_key: 'author_id'
+
```
Comment (DeliceLydia) [inline app/models/user.rb]: - [ ] A user should have many `comments` and `likes` as well

File: app/models/post.rb [hunk 2]
```diff
@@ -1,3 +1,15 @@
 class Post < ApplicationRecord
-  belongs_to :user
+  belongs_to :user, foreign_key: 'author_id'
+  has_many :likes, foreign_key: 'post_id'
+  has_many :comments, foreign_key: 'post_id'
+
+  def self.update_post_counter(user_id)
+    user = User.find_by(id: user_id)
+    user.update(posts_counter: user.posts.count)
+  end
```
Comment (DeliceLydia) [inline app/models/post.rb]: ## Optional
- [ ] I suggest that you use one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the `#update_post_counter` method every time a new post has been saved to the database, then make the `update_post_counter` method private

File: app/models/like.rb [hunk 3]
```diff
@@ -1,4 +1,9 @@
 class Like < ApplicationRecord
   belongs_to :post
   belongs_to :user
+
+  def self.update_like_counter(post_id)
+    post = Post.find_by(id: post_id)
+    post.update(likes_counter: post.likes.count)
+  end
```
Comment (DeliceLydia) [inline app/models/like.rb]: ## Optional
- [ ] I suggest that you use one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the `#update_like_counter` method every time a new like has been saved to the database, then make the `update_like_counter` method private

File: app/models/comment.rb [hunk 4]
```diff
@@ -1,4 +1,9 @@
 class Comment < ApplicationRecord
   belongs_to :post
   belongs_to :user
+
+  def self.update_comment_counter(post_id)
+    post = Post.find_by(id: post_id)
+    post.update(comments_counter: post.comments.count)
+  end
```
Comment (DeliceLydia) [inline app/models/comment.rb]: ## Optional
- [ ] I suggest that you use one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the `#update_comment_counter` method every time a new comment has been saved to the database, then make the `update_comment_counter` method private

**Review by DeliceLydia — CHANGES_REQUESTED** [review]
## Hi @batoolfatima2135, 

Good job so far! There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- The user model has a method that returns 3 most recent posts.
- Post model has a method that updates the posts counter for a user
- The post model has a method that returns the 5 most recent comments for a given post
- Followed gitflow


## Required Changes ♻️
_Check the comments under the review._



### Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by mukhammadrizooff — APPROVED** [review]
Hi @batoolfatima2135, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉

 

## Highlights
- No linter errors ✔️ 
- Correct GitFlow ✔️ 
- Good implementations ✔️
- Descriptive `README` file ✔️

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**